### ISSUE TITLE
No error message displayed when attempting to report abuse while another abuse report instance is open

### ISSUE THREAD

**Issue by AlexandraMoga** [mozilla/addons-frontend#8988]
Prerequisites: make sure that your browser has all the necessary prefs mentioned in https://github.com/mozilla/addons-server/issues/12741

### Describe the problem and steps to reproduce it:

1. Load AMO -dev in a Firefox version that supports the new mozAddonManager.reportAbuse method
2. Open an add-on details page 
3. Click on Report this add-on for abuse button
4. After the report abuse overlay is opened, click anywhere outside the overlay - this should minimize the abuse report form
5. Refresh the add-on details page
6. Click again on the Report this add-on for abuse button

### What happened?
Nothing happens when the Report this add-on for abuse button is clicked the second time

### What did you expect to happen?
Show an error message in the AMO report abuse component to notify the user that a report abuse instance is already opened

### Anything else we should know?
- the console will respond with the following error message when the button is clicked the second time:
`Reporting add-on for abuse via firefox failed: Error: An abuse report is already in progress`
- this scenario could happen when users accidentally minimize the abuse report form and, without noticing that the frame is still opened in the taskbar, they might be tempted to re-start the report abuse process
- reproduced on AMO -dev with Fx 71 and 72, Win10x64

![2019-11-26_15-49-39](https://user-images.githubusercontent.com/31961530/69640121-7ee70780-1066-11ea-8a93-d0c1231acf80.gif)

**Comment by bobsilverberg** [comment]
@rpl We discussed this a bit in our weekly standup and we think that perhaps this needs to be dealt with on the Firefox side. Here's the scenario:

If one initiates an Abuse Report from AMO, and one clicks away from the Report Abuse dialog, the dialog gets hidden from the user, but the abuse report is considered in progress. If a user then refreshes the browser and tries to report abuse again, for the same add-on, or even for a different add-on, the API returns an error saying "Reporting add-on for abuse via firefox failed: Error: An abuse report is already in progress". 

The problem is that there's not much that we can do from the AMO side at this point. The dialog is hidden. It can be accessed from the Window menu, but I'm not sure there's anything that AMO can do to refocus the window or help the user find it.

It seems like a better solution might be to cancel the in-progress abuse report if a second one is started and allow the second one to proceed. There may be other ways of approaching it as well, and I'd be keen to hear what you think we can do to address this problem.

**Comment by willdurand** [comment]
> [...] or even for a different add-on [...]

The STR in the OP are quite involved (loosing the popup focus, fully reloading the page, clicking the report button again) but that part is very much plausible.

**Comment by bobsilverberg** [comment]
> > [...] or even for a different add-on [...]
> 
> The STR in the OP are quite involved (loosing the popup focus, fully reloading the page, clicking the report button again) but that part is very much plausible.

You actually still need to refresh the page, even if you navigate to a different add-on, as the redux store also keeps track of an abuse report in progress and disables the button until that report finishes. While that's good for avoiding the error, it doesn't help a user understand *why* they cannot report abuse, so this is still a problem in this situation, and is in fact a bit different.

@rpl Does the promise ever "time out" if a user does not complete the Report Abuse flow via the Firefox dialog? Was using a modal dialog considered, as this would eliminate this issue entirely, but I guess we might not want to interfere with users' other tabs.

**Comment by rpl** [comment]
@bobsilverberg apologies for totally missing the ping yesterday

> @rpl Does the promise ever "time out" if a user does not complete the Report Abuse flow via the Firefox dialog?

No, there is no timeout at the moment (and the existing dialog window would still be able to submit a report even if the AMO tab that triggered the report has been closed).

I totally agree that if anything has to be done to cover this scenario, it has to be done on the Firefox side.

Some simplest options to deal with it could be:

- option 1: still reject the new reportAbuse calls, but also bring any the existing dialog window on top
- option 2: close the previous dialog window and open a new one for the last reportAbuse call
- option 3:  like option 2 + option 1:
  - close the previous dialog window only if the user hasn't yet changed anything in it
  - and reject the new call (and bring the old dialog window on top) if the user did change anything in the old one

>  Was using a modal dialog considered, as this would eliminate this issue entirely, but I guess we might not want to interfere with users' other tabs.

In my initial prototypes I did try both a modal and a non modal dialog, if the dialog is modal the Firefox window that has triggered the modal dialog is blocked and so if the user clicks on a link part of the report panel (e.g. one of the "learn more" links) we have to open a new separate window.

(Besides that, a modal dialog window on osx has a totally different semi-transparent styling, different enough to actually require some different styling and maybe ux choices, see https://bugzilla.mozilla.org/attachment.cgi?id=9088166).

In the end, the behavior that we agreed on was a non-modal dialog "alwaysOnTop", as it did seem as a reasonable way to reduce the chances that the user may ever lose track of the dialog window, and then the panel links can be opened in the same Firefox window that triggered the dialog (as it isn't blocked and it can still be used by the user).

We ended up to remove the alwaysOnTop as part of [Bug 1595091](https://bugzilla.mozilla.org/show_bug.cgi?id=1595091) and uplifted the change, which I guess is what has (kind of) made this issue "possible".

**Comment by bobsilverberg** [comment]
> 
> Some simplest options to deal with it could be:
> 
>     * option 1: still reject the new reportAbuse calls, but also bring any the existing dialog window on top
> 
>     * option 2: close the previous dialog window and open a new one for the last reportAbuse call
> 
>     * option 3:  like option 2 + option 1:
>       
>       * close the previous dialog window only if the user hasn't yet changed anything in it
>       * and reject the new call (and bring the old dialog window on top) if the user did change anything in the old one
>

I think option 2 is best. I understand the concern trying to be addressed with options 1 and 3, but I think that could be confusing for a user if they have now clicked on a *different* add-on. Maybe an option 4 would be similar to option 3, but also check the addonId being reported, and only re-focus the new window if the addonId being submitted is the same as the one in the existing window.

@rpl What do you think?

**Comment by rpl** [comment]
@bobsilverberg yeah, I don't disagree, option 1 and option 3 (if implemented as described) may still be confusing.

option 2 sounds good to me too, I'm going to file an issue and looking into that. 
(option 4 also looks something that we could implemented without too many changes, I'll also take a look into that).

**Comment by bobsilverberg** [comment]
Thanks @rpl 

@jvillalobos As this is going to be addressed on the Firefox side, it's not something that we can necessarily fix quickly. Do we want to consider disabling this flow until this issue is addressed? I don't imagine a large number of people will encounter this, so I'm not sure that we need to do that, but I figured it prudent to ask.

**Comment by jvillalobos** [comment]
No, it seems like a very unlikely problem to run into, so we can live with that bug for now.

**Comment by rpl** [comment]
I forgot to mention here that I filed [Bug 1602758 - Close an existing abuse report dialog instead of rejecting a new mozAddonManager.reportAbuse](https://bugzilla.mozilla.org/show_bug.cgi?id=1602758) and attached a patch to it.

I did push it to autoland today, and so it should likely be there in the next nightly build.

**Comment by bobsilverberg** [comment]
I am going to close this, as it was fixed via https://bugzilla.mozilla.org/show_bug.cgi?id=1602758. It would be a good idea to verify it as fixed, so I'm going to assign this week's milestone to it, even though it wasn't fixed by any code that we landed. If there is a better way to make sure that this is verified please let me know.

There is another issue related to this, which I have opened at https://github.com/mozilla/addons-frontend/issues/9086.

**Comment by AlexandraMoga** [comment]
I've verified the above scenario with Beta 73 with the fix for #9086 and the issue looks fixed.
If the report abuse dialog is closed by accident or if the user navigates to another add-on details page without completing the previously opened abuse report, the old dialog will be closed automatically and a new dialog (with the new add-on details) will open when the users clicks on the 'Report this add-on for abuse' button:

![2020-01-14_16-35-52](https://user-images.githubusercontent.com/31961530/72354039-b761a400-36ed-11ea-9239-8419b6709ffa.gif)

_Note that, since the fix for this enters only from Fx 73, the scenario will continue to reproduce in the current release._
